cloudant-http/src/main/java/com/cloudant/http/interceptors/CookieInterceptor.java

```diff
@@ -79,86 +84,123 @@ public HttpConnectionInterceptorContext interceptRequest(HttpConnectionIntercept
+                                            HttpConnectionInterceptorException(errorString);
+                                    // Set the flag for deserialization
+                                    toThrow.deserialize = true;
+                                    throw toThrow;
+                                } else {
+                                    // Was expired - set boolean to renew cookie
+                                    renewCookie = true;
+                                }
+                                break;
+                            case HttpURLConnection.HTTP_UNAUTHORIZED: //401
+                                IOUtils.toString(errorStream, "UTF-8");
+                                // We need to get a new cookie
                                 renewCookie = true;
-                            }
-                            break;
-                        case HttpURLConnection.HTTP_UNAUTHORIZED: //401
-                            IOUtils.toString(errorStream, "UTF-8");
-                            // We need to get a new cookie
-                            renewCookie = true;
-                            break;
-                        default:
-                            break;
+                                break;
+                            default:
+                                break;
+                        }
+                    } finally {
+                        IOUtils.closeQuietly(errorStream);
+                    }
+                    if (renewCookie) {
+                        logger.finest("Cookie was invalid attempt to get new cookie.");
+                        boolean success = requestCookie(connection.getURL(), context);
+                        if (success) {
+                            // New cookie obtained, replay the request
+                            context.replayRequest = true;
+                        } else {
+                            // Didn't successfully renew, maybe creds are invalid
+                            context.replayRequest = false; // Don't replay
+                            shouldAttemptCookieRequest.set(false); // Set the flag to stop trying
+                        }
                     }
-                } finally {
-                    IOUtils.closeQuietly(errorStream);
-                }
-            }
-
-            if (renewCookie) {
-                cookie = getCookie(connection.getURL(), context);
-                // Don't resend request, failed to get cookie
-                if (cookie != null) {
-                    context.replayRequest = true;
                 } else {
-                    context.replayRequest = false;
+                    // Store any cookies provided on the response
+                    storeCookiesFromResponse(connection);
                 }
+            } catch (IOException e) {
+                logger.log(Level.SEVERE, "Failed to get response code from request", e);
```

Comment:
It's possible we could have got here because we failed to read the error stream as well - I don't know if that makes the failure too verbose